File: 백준 12851 BFS 숨바꼭질2.cs
using System;
using System.Collections.Generic;

class Program
{
    const int MAX = 100001;

    static void Main(string[] args)
    {
        string[] inputs = Console.ReadLine().Split(' ');
        int N = int.Parse(inputs[0]);
        int K = int.Parse(inputs[1]);

        int[] visited = new int[MAX];
        Queue<int> queue = new Queue<int>();
        queue.Enqueue(N);

        int cnt = 0;
        int result = 0;

        while (queue.Count > 0)
        {
            int curr = queue.Dequeue();

            if (curr == K)
            {
                if (result == 0) result = visited[curr];
                if (result == visited[curr]) cnt++;
                continue;
            }

            int[] nextNums = { curr - 1, curr + 1, curr * 2 };
            foreach (int nextNum in nextNums)
            {
                if (0 <= nextNum && nextNum < MAX && (visited[nextNum] == 0 || visited[nextNum] == visited[curr] + 1))
                {
                    visited[nextNum] = visited[curr] + 1;
                    queue.Enqueue(nextNum);
                }
            }
        }

        Console.WriteLine(result);
        Console.WriteLine(cnt);
    }
}

File: 백준 12851 BFS 숨바꼭질2.py
from collections import deque
import sys

input = sys.stdin.readline
N, K = map(int, input().split())
queue = deque()
queue.append(N)
visited = [0] * 100001
cnt, result = 0, 0
while queue:
    curr = queue.popleft()
    temp = visited[curr]
    if curr == K:
        result = temp
        cnt += 1
        continue

    for nxt_num in [curr - 1, curr + 1, curr * 2]:
        if 0 <= nxt_num < 100001 and (
            visited[nxt_num] == 0 or visited[nxt_num] == visited[curr] + 1
        ):
            visited[nxt_num] = visited[curr] + 1
            queue.append(nxt_num)
print(result)
print(cnt)
